aten/src/ATen/native/SobolEngineOps.cpp

```diff
@@ -0,0 +1,141 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+#include "ATen/native/SobolEngineOpsUtils.h"
+
+#include <vector>
+
+namespace at {
+namespace native {
+
+/// This is the core function to draw samples from a `SobolEngine` given
+/// its state variables (`sobolstate` and `quasi`). `dimension` can be
+/// inferred from `sobolstate`, but choosing to pass it explicitly to avoid
+/// an extra operation to obtain the size of the first dimension of
+/// `sobolstate`.
+std::tuple<Tensor, Tensor> _sobol_engine_draw(const Tensor& quasi, int64_t n, const Tensor& sobolstate,
+                                              int64_t dimension, int64_t num_generated) {
+  Tensor wquasi = quasi.clone();
+  std::vector<Tensor> result;
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+  AT_CHECK(wquasi.type().scalarType() == at::ScalarType::Long,
+           "quasi needs to be of type ", at::ScalarType::Long);
+
+  for (int64_t i = 0; i < n; ++i) {
+    int64_t l = rightmost_zero(num_generated);
+    Tensor inter_res = wquasi.__ixor__(sobolstate.select(1, l - 1));
+    result.emplace_back(inter_res.clone());
+    num_generated++;
+  }
+
+  return std::make_tuple(at::native::stack(result, 0).toType(at::kFloat).mul_(RECIPD), wquasi);
+}
+
+/// This is the core function to fast-forward a `SobolEngine` given
+/// its state variables (`sobolstate` and `quasi`). `dimension` can be
+/// inferred from `sobolstate`, but is passed as an argument for the same reasons
+/// specified above.
+Tensor _sobol_engine_ff(const Tensor& quasi, int64_t n, const Tensor& sobolstate,
+                        int64_t dimension, int64_t num_generated) {
+  Tensor wquasi = quasi.clone();
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+  AT_CHECK(wquasi.type().scalarType() == at::ScalarType::Long,
+           "quasi needs to be of type ", at::ScalarType::Long);
+
+  for (int64_t i = 0; i < n; ++i) {
+    int64_t l = rightmost_zero(num_generated);
+    wquasi.__ixor__(sobolstate.select(1, l - 1));
+    num_generated++;
+  }
+  return wquasi;
+}
+
+/// This is an implicit function used for randomizing the state variables of the.
+/// `SobolEngine`. Arguments are a randomized `sobolstate` state variables
+/// and a list of random lower triangular matrices consisting of 0s and 1s. `dimension` is
+/// passed explicitly again.
+Tensor _sobol_engine_scramble(const Tensor& sobolstate, TensorList ltm, int64_t dimension) {
```

Comment:
There seems to be something wrong in this implementation that I am not able to figure out. Some help will be appreciated :).